### PR TITLE
save keras version & compile args when serializing models

### LINKED ISSUE #2498: save optimizer and objective information to configure file
It seems that keras 1 doesn't support save optimizer and loss to configure file any more. It there some alternative way? Or any plan to add this function back again?

### PR DESCRIPTION
**Pull request by cmyr**
I frequently load and retrain models serialized weeks and months earlier, and it's useful to know the `optimizer` and `loss` + other args that a model was compiled with. Other people have mentioned this as well.

Thought this was simple enough that I'd take a stab. This doesn't effect any existing functionality, but it lets me manually get the required information out of a json/yaml file.

I also took the liberty of saving the keras version number, which was mentioned in a `TODO` comment.

if @fchollet or anyone else has a better idea for how to do this (or why not to do this) I'm happy to revisit.

(fixes #2498)

edit: failing a pep8 test, I'll take a look in a bit and push again

### CODE REVIEW

File: keras/engine/topology.py [hunk 1]
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
             K.batch_set_value(weight_value_tuples)
         f.close()
 
+    def prepare_config(self):
```
Comment (fchollet) [inline keras/engine/topology.py]: I find that having `prepare_config` and `get_config` could be confusing... Maybe better naming would help?

File: keras/engine/topology.py [hunk 2]
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
...
+
+        config = self.get_config()
+        model_config = {
+            'class_name': self.__class__.__name__,
+            'config': config,
+            'keras_version': keras_version
+        }
+
+        if hasattr(self, 'optimizer'):
+            model_config['optimizer'] = prepare(self.optimizer)
```
Comment (fchollet) [inline keras/engine/topology.py]: Optimizers have a `get_config` method precisely for this purpose.

File: keras/engine/topology.py [hunk 3]
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
...
+            model_config['loss'] = prepare(self.loss)
```
Comment (fchollet) [inline keras/engine/topology.py]: `__class__.__name__` should suffice here?

File: keras/engine/topology.py [hunk 4]
```diff
@@ -2348,6 +2346,41 @@ def load_weights(self, filepath):
             K.batch_set_value(weight_value_tuples)
         f.close()
 
+    def prepare_config(self):
+        '''shared between different serialization methods'''
+        from six import string_types
+        from keras import __version__ as keras_version
+
+        def prepare(obj):
```
Comment (fchollet) [inline keras/engine/topology.py]: Not sure this is needed.

File: keras/engine/topology.py [hunk 5]
```diff
@@ -2367,11 +2400,7 @@ def get_json_type(obj):
 
             raise TypeError('Not JSON Serializable')
 
-        config = self.get_config()
-        model_config = {
-            'class_name': self.__class__.__name__,
-            'config': config,
-        }
+        model_config = self.prepare_config()
```
Comment (fchollet) [inline keras/engine/topology.py]: Sufficiently trimmed, `prepare_config` could fit in 5 lines, so maybe it's not so bad to have it twice? If factored out, it should have a better name anyway.

**Comment by cmyr** [comment]
@fchollet alright, addressed some of your comments. Not sure if using the underbar for a name really fits code-style wise, but my imagination let me down. Happy to change if you have another idea.

**Comment by fchollet** [comment]
> Not sure if using the underbar for a name really fits code-style wise, but my imagination let me down.

It's reasonable, it's used in many other parts of the code to indicate internal methods.

LGTM

**Pull request merged**